### DISCUSSION TITLE
Functionality of the colormap plot style.

### DISCUSSION THREAD

**Discussion by Maxime-CS** [PinkShnack/TEMUL#70] [Q&A] [answered]
Hi,
I am starting to discover Temul and trying to plot polarization vectors.
Is it possible to control the scale bar (magnitude of arrows) that is given when we use the colormap plot style? I would like to fix the min and the max value for an easy comparison of different areas. 
Sorry for this question if it is already implemented in the code. 
Best
Maxime

**Comment by PinkShnack** [comment]
Hey!

You can change the arrow size within the `plot_polarisation_vectors` function by using some of its parameters. The function parameter descriptions are described here in the docs: https://temul-toolkit.readthedocs.io/en/latest/api_doc.html#temul.topotem.polarisation.plot_polarisation_vectors

Some example parameters that you could change would be:
 - quiver_units
 - pivot
 - angles
 - scale_units
 - scale
 - headwidth
 - headlength
 - headaxislength
 - width
 - minshaft
 - minlength

These are not described in our docs because (I was lazy) and because they are taken directly from the matplotlib quiver function here: https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.quiver.html. There you will see parameter descriptions to change the arrow sizes/length etc.

In your specific case, the `units` parameter (called `quiver_units` in temul) is what you might be looking for. So setting `quiver_units='dots'` might be a good starting point. Furthermore `scale` and `scale_units` might be useful to play around with.

The matplotlib quiver function parameters is a bit tricky to use, and it is something that sometimes requires you to mess around with a lot to get the values you want. Just to clarify, you just change the above parameters when running the `plot_polarisation_vectors` function. You don't need to call the quiver function yourself!

Questions are always welcome! As you can see, it is implemented already, but the answer is not so straightforward.

**Reply by PinkShnack** [reply]
If the above answer didn't help, please elaborate your question. :)

**Comment by Maxime-CS** [comment]
Hi ! 
Many thanks for your anwser and all these advises.
My question was more about the possibility to manipulate the colorbar given in the colomap plot style. In the following image, I would like to set the maximal value corresponding to the yellow to 0.05 nm instead of being dependant of the amplitude found in the image (here 0.037 nm). The same goes for the minimum value but I guess that the "minlength" option can help to fix it. 

![Figure_12](https://user-images.githubusercontent.com/102516161/161727524-cf24fe1d-3644-4bb2-b551-22165e005fa3.jpeg)

**Reply by PinkShnack** [reply]
Ah I see what you mean now. I can add this functionality soon, maybe this weekend! (or you are welcome to create a pull request and make the changes yourself).

- Here is where one could use optional parameters, maybe called `cbar_vmin` and `cbar_vmax`: https://github.com/PinkShnack/TEMUL/blob/master/temul/topotem/polarisation.py#L437

**Reply by PinkShnack** [reply]
It might also be an option to just allow `**kwargs` for cbar. Would be nice to have kwargs that for a lot of function calls within our function!

**Reply by PinkShnack** [reply]
If you're eager to try the new functionality, you can  by installing the `temul-toolkit` at the new branch for now. Use the following command to install via pip (I unfortunately haven't checked if this works yet!

```bash
pip install temul-toolkit@git+https://github.com/PinkShnack/TEMUL.git@add_cbar_params
```

Then you can use the new params to set the min, max or both for the colorbar! `cbar_vmin` and `cbar_max`. For example:

```python

cbar_vmin = 0.2
cbar_vmax = 1.0

tml.plot_polarisation_vectors(
        x, y, u, v, image=sublatticeA, save=None,
        plot_style='colormap', vector_rep=vector_rep,
        cbar_vmin=cbar_vmin, cbar_vmax=cbar_vmax)
```

**Reply by PinkShnack** [reply]
Let me know if this would solve your problem @Maxime-CS.

Because the function is getting a bit ridiculous (with so many parameters), and isn't well designed, I will likely try to refactor it into a class that is much more flexible and extensible, and not limited to parameters. See #74 for details.

**Reply by Maxime-CS** [reply]
Hi PinkShnack,
I finally got time to try it. It's not working. Perhaps I miss something...
I checked that I had the 0.1.7 of temul-toolkit then I write the command you send me ("pip install temul-toolkit@git+https://github.com/PinkShnack/TEMUL.git@add_cbar_params").
Here is what I got when I use cbar_vmin/max

![Image2](https://user-images.githubusercontent.com/102516161/165716873-bbee7246-32b3-4085-95aa-44b6dbd5708b.jpg)

If I use "TAB" on the function :
- I can find cbar option in the beginning of the function
- but not after the "plot_style". See image

![Image1](https://user-images.githubusercontent.com/102516161/165716463-cd5dd526-ff82-4346-94c9-53e7f72dc0da.jpg)

Do I miss something ? Anyway I would like to thank you for your time and quick responses. 
Best

**Reply by PinkShnack** [reply]
Hey, sorry for the late reply. First we'd better check if your installed version of `temul` is correct. You ran the [pip command from above](https://github.com/PinkShnack/TEMUL/discussions/70#discussioncomment-2526141), right?
- Open your terminal (prompt) and activate the environment you normally use for `temul`. If you don't work with environments, then don't worry.
- Just run `pip list` and scroll up to where `temul-toolkit` is. It will give you a version. It should be something like `0.1.7.post34`. Let me know what your version is.

EDIT: the problem could also be an install issue with pip not allowing you to install directly a https link without admin rights. So please also paste the output of the `pip install temul-toolkit@git+https://github.com/PinkShnack/TEMUL.git@add_cbar_params` command.

**Reply by Maxime-CS** [reply]
Hi, with the help of a collegue, I finally get the branch version. It works now. Thanks a lot. (I was on the master branch 0.1.7 instead of add_cbar_params one and it was not so obvious to change. 
![temul](https://user-images.githubusercontent.com/102516161/167380302-1b68c528-a178-4b98-ae59-25b46da3ac50.png)

**Reply by PinkShnack** [reply]
Yeah these things can be tricky! Good to know it works. I will merge the changes soon and release a new version. Then the install will be easier! If you have any feedback while using it, please comment here!

**Comment by PinkShnack** [comment, marked as answer]
For the current working answer to the question, see this comment (also pasted below): https://github.com/PinkShnack/TEMUL/discussions/70#discussioncomment-2526141. When a new version is released, the install will be done as normal with pip.

> If you're eager to try the new functionality, you can by installing the `temul-toolkit` at the new branch for now. Use the following command to install via pip (I unfortunately haven't checked if this works yet!
> 
> ```shell
> pip install temul-toolkit@git+https://github.com/PinkShnack/TEMUL.git@add_cbar_params
> ```
> 
> Then you can use the new params to set the min, max or both for the colorbar! `cbar_vmin` and `cbar_max`. For example:
> 
> ```python
> cbar_vmin = 0.2
> cbar_vmax = 1.0
> 
> tml.plot_polarisation_vectors(
>         x, y, u, v, image=sublatticeA, save=None,
>         plot_style='colormap', vector_rep=vector_rep,
>         cbar_vmin=cbar_vmin, cbar_vmax=cbar_vmax)
> ```

As noted in https://github.com/PinkShnack/TEMUL/discussions/70#discussioncomment-2539224, this function probably needs a rework. See #74 for details.

**Comment by PinkShnack** [comment]
New version from 0.1.9 will have this fixed.
